google/cloud/bigtable/data/_metrics/data_model.py

```diff
@@ -395,7 +403,17 @@ def wrapped_predicate(exc: Exception) -> bool:
                 self.end_with_status(exc)
             return inner_result
 
-        return wrapped_predicate
+        def wrapped_exception_factory(
+            exc_list: list[Exception],
+            reason: RetryFailureReason,
+            timeout_val: float | None,
+        ) -> tuple[Exception, Exception | None]:
+            exc, source = _retry_exception_factory(exc_list, reason, timeout_val)
+            if reason != RetryFailureReason.NON_RETRYABLE_ERROR:
```

Comment:
should this be if reason = RetryFailureReason.NON_RETRYABLE_ERROR?

Reply:
No, the `wrapped_predicate` will report all exceptions encountered. But the predicate isn't called when the operation ends due to timeout. So this extra wrapper is needed for that

This part is a bit more complicated than I hoped, so I was considering trying to refactor some of the read_rows instrumentation when I have a chance. Let me know what you think

Reply:
I see. It makes sense but a little hard to follow. Maybe add a comment for now?  And also in _read_rows start_operation  comment what `metric_fns[0]` and `metric_fns[1]` are?